File: packages/core/src/api-client.ts
import { sessionFromToken } from "./session";
import type { CameraSession, CaptureResponse } from "./types";

export class ApiClient {
  constructor(
    private readonly baseUrl: string,
    private token: string,
  ) {}

  /** Exchange an OTP for a JWT session. */
  async connect(otp: string): Promise<CameraSession> {
    const res = await fetch(`${this.baseUrl}/camera/connect`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ otp: otp.toUpperCase() }),
    });
    if (!res.ok) {
      const data = await res.json().catch(() => ({})) as { error?: string };
      throw new Error(data.error ?? `Connect failed: ${res.status}`);
    }
    const { token } = await res.json() as { token: string };
    this.token = token;
    return sessionFromToken(token);
  }

  /** Send a batch of face images. Returns matched persons. */
  async capture(images: string[]): Promise<CaptureResponse> {
    const res = await fetch(`${this.baseUrl}/camera/capture`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.token}`,
      },
      body: JSON.stringify({ images }),
    });
    if (!res.ok) throw new Error(`Capture failed: ${res.status}`);
    return res.json() as Promise<CaptureResponse>;
  }

  /** Heartbeat — validate the camera session is still active. */
  async heartbeat(): Promise<{ id: string; name: string; type: string }> {
    const res = await fetch(`${this.baseUrl}/camera/me`, {
      headers: { Authorization: `Bearer ${this.token}` },
    });
    if (!res.ok) throw Object.assign(new Error("Heartbeat failed"), { status: res.status });
    return res.json() as Promise<{ id: string; name: string; type: string }>;
  }
}

File: packages/flutter/lib/src/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'session.dart';
import 'types.dart';

class ApiClient {
  ApiClient({required this.baseUrl, required this.token});

  final String baseUrl;
  String token;

  /// Exchange a 6-char OTP for a JWT session.
  Future<CameraSession> connect(String otp) async {
    final res = await http.post(
      Uri.parse('$baseUrl/camera/connect'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({'otp': otp.toUpperCase()}),
    );
    if (res.statusCode != 200 && res.statusCode != 201) {
      final body = json.decode(res.body) as Map<String, dynamic>;
      throw Exception(body['error'] ?? 'Connect failed: ${res.statusCode}');
    }
    final body    = json.decode(res.body) as Map<String, dynamic>;
    token         = body['token'] as String;
    final session = sessionFromToken(token);
    return session;
  }

  /// Send a batch of face images. Returns the list of matched persons.
  Future<List<PersonResult>> capture(List<String> images) async {
    final res = await http.post(
      Uri.parse('$baseUrl/camera/capture'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: json.encode({'images': images}),
    );
    if (res.statusCode != 200) {
      throw _ApiException(res.statusCode, 'Capture failed');
    }
    final body    = json.decode(res.body) as Map<String, dynamic>;
    final persons = (body['persons'] as List<dynamic>)
        .map((p) => PersonResult.fromJson(p as Map<String, dynamic>))
        .toList();
    return persons;
  }

  /// Heartbeat — verify the session is still active.
  Future<Map<String, dynamic>> heartbeat() async {
    final res = await http.get(
      Uri.parse('$baseUrl/camera/me'),
      headers: {'Authorization': 'Bearer $token'},
    );
    if (res.statusCode != 200) {
      throw _ApiException(res.statusCode, 'Heartbeat failed');
    }
    return json.decode(res.body) as Map<String, dynamic>;
  }
}

class _ApiException implements Exception {
  _ApiException(this.statusCode, this.message);
  final int statusCode;
  final String message;
  @override
  String toString() => 'ApiException($statusCode): $message';
}
